### ISSUE TITLE
BINDetect p-value issue

### ISSUE THREAD

**Issue by Danvor** [loosolab/TOBIAS#15]
Hi, I have an issue when running the BINDetect. I get this issue at default p-value and around that, however if I set a p-value cut-off sufficiently high, e.g. 5e-2, I dont have the error and everything works fine. 

See the commands and error messages below:
```

# TOBIAS 0.11.1-dev BINDetect (run started 2020-05-04 16:08:06.715221)
# Working directory: /home/danvor/Data/
# Command line call: TOBIAS BINDetect --signals ATAC_footprints.bw --motifs /home/danvor/Scripts/jaspar2020_vertebrate.jaspar --genome /home/danvor/Data/genome.fa.gz --peaks /home/danvor/Data/GOI_CRMs.bed --outdir GOI_motifs/ --prefix GOI_TF --motif-pvalue 1e-3 --cores 16 --naming name --cond_names GOI --skip-excel

# ----- Input parameters -----
# signals:	['ATAC_footprints.bw']
# peaks:	/home/danvor/Data/GOI_CRMs.bed
# motifs:	['/home/danvor/Scripts/jaspar2020_vertebrate.jaspar']
# genome:	/home/danvor/Data/genome.fa.gz
# cond_names:	['GOI']
# peak_header:	None
# naming:	name
# motif_pvalue:	0.001
# bound_pvalue:	0.001
# pseudo:	None
# time_series:	False
# skip_excel:	True
# prefix:	GOI_TF
# outdir:	/home/danvor/Data/GOI_motifs
# cores:	16
# split:	100
# verbosity:	3


# ----- Output files -----
# /home/danvor/Data/GOI_motifs/*/beds/*_GOI_bound.bed
# /home/danvor/Data/GOI_motifs/*/beds/*_GOI_unbound.bed
# /home/danvor/Data/GOI_motifs/*/beds/*_all.bed
# /home/danvor/Data/GOI_motifs/*/plots/*_log2fcs.pdf
# /home/danvor/Data/GOI_motifs/*/*_overview.txt
# /home/danvor/Data/GOI_motifs/*/*_overview.xlsx
# /home/danvor/Data/GOI_motifs/GOI_TF_distances.txt
# /home/danvor/Data/GOI_motifs/GOI_TF_results.txt
# /home/danvor/Data/GOI_motifs/GOI_TF_results.xlsx
# /home/danvor/Data/GOI_motifs/GOI_TF_figures.pdf


/opt/pyenv/tobias-env/lib/python3.8/site-packages/logomaker-0.8-py3.8.egg/logomaker/src/validate.py:98: SyntaxWarning: "is" with a literal. Did you mean "=="?
  if matrix_type is 'information':
/opt/pyenv/tobias-env/lib/python3.8/site-packages/logomaker-0.8-py3.8.egg/logomaker/src/validate.py:104: SyntaxWarning: "is" with a literal. Did you mean "=="?
  elif matrix_type is 'probability':
2020-05-04 16:08:06 (1329) [INFO]	----- Processing input data -----
2020-05-04 16:08:06 (1329) [INFO]	Checking reading/writing of files
2020-05-04 16:08:06 (1329) [INFO]	Reading peaks
2020-05-04 16:08:06 (1329) [INFO]	- Found 7 regions in input peaks
2020-05-04 16:08:06 (1329) [INFO]	- Merged to 7 regions
2020-05-04 16:08:06 (1329) [INFO]	Checking for match between --peaks and --fasta/--signals boundaries
2020-05-04 16:08:06 (1329) [INFO]	- Comparing peaks to /home/danvor/Data/genome.fa.gz
2020-05-04 16:08:06 (1329) [INFO]	- Comparing peaks to ATAC_footprints.bw
2020-05-04 16:08:07 (1329) [INFO]	Estimating GC content from peak sequences
2020-05-04 16:08:07 (1329) [INFO]	- GC content estimated at 45.85%
2020-05-04 16:08:07 (1329) [INFO]	Reading motifs from file
2020-05-04 16:08:07 (1329) [INFO]	- Read 746 motifs
2020-05-04 16:08:09 (1329) [INFO]	Creating folder structure for each TF
2020-05-04 16:08:25 (1329) [INFO]	Plotting sequence logos for each motif
2020-05-04 16:08:25 (1329) [INFO]	Progress 0%
2020-05-04 16:08:26 (1329) [INFO]	Progress 1.34%
2020-05-04 16:08:26 (1329) [INFO]	Progress 2.01%
2020-05-04 16:08:26 (1329) [INFO]	Progress 2.28%
...
2020-05-04 16:08:39 (1329) [INFO]	Progress 100.0%
2020-05-04 16:08:39 (1329) [INFO]	Progress done!


2020-05-04 16:08:40 (1329) [INFO]	Scanning for motifs and matching to signals...
2020-05-04 16:08:40 (1329) [INFO]	Progress 0%
2020-05-04 16:08:43 (1329) [INFO]	Progress 29.0%
...
2020-05-04 16:08:44 (1329) [INFO]	Progress 100.0%
2020-05-04 16:08:44 (1329) [INFO]	Progress done!
2020-05-04 16:08:44 (1329) [INFO]	Done scanning for TFBS across regions!
2020-05-04 16:08:44 (1329) [INFO]	Waiting for bedfiles to write
2020-05-04 16:08:46 (1329) [INFO]	Merging results from subsets

2020-05-04 16:08:46 (1329) [INFO]	Estimating score distribution per condition
2020-05-04 16:08:47 (1329) [INFO]	Normalizing scores
2020-05-04 16:08:47 (1329) [INFO]	Estimating bound/unbound threshold
2020-05-04 16:08:47 (1329) [STATS]	- Threshold estimated at: 3.53463


2020-05-04 16:08:47 (1329) [INFO]	Processing scanned TFBS individually
2020-05-04 16:08:47 (1329) [INFO]	Progress 0%
2020-05-04 16:08:47 (1329) [INFO]	Progress 1.07%
...
2020-05-04 16:08:52 (1329) [INFO]	Progress 100.0%
2020-05-04 16:08:52 (1329) [INFO]	Progress done!
multiprocessing.pool.RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/opt/pyenv/tobias-env/lib/python3.8/site-packages/pandas/core/indexes/base.py", line 2646, in get_loc
    return self._engine.get_loc(key)
  File "pandas/_libs/index.pyx", line 111, in pandas._libs.index.IndexEngine.get_loc
  File "pandas/_libs/index.pyx", line 138, in pandas._libs.index.IndexEngine.get_loc
  File "pandas/_libs/hashtable_class_helper.pxi", line 1619, in pandas._libs.hashtable.PyObjectHashTable.get_item
  File "pandas/_libs/hashtable_class_helper.pxi", line 1627, in pandas._libs.hashtable.PyObjectHashTable.get_item
KeyError: 'GOI_score'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/python-3.8.2/lib/python3.8/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/opt/pyenv/tobias-env/lib/python3.8/site-packages/tobias-0.11.1.dev0-py3.8-linux-x86_64.egg/tobias/tools/bindetect_functions.py", line 370, in process_tfbs
    info_table.at[TF_name, condition + "_mean_score"] = round(np.mean(bed_table[condition + "_score"]), 5)
  File "/opt/pyenv/tobias-env/lib/python3.8/site-packages/pandas/core/frame.py", line 2800, in __getitem__
    indexer = self.columns.get_loc(key)
  File "/opt/pyenv/tobias-env/lib/python3.8/site-packages/pandas/core/indexes/base.py", line 2648, in get_loc
    return self._engine.get_loc(self._maybe_cast_indexer(key))
  File "pandas/_libs/index.pyx", line 111, in pandas._libs.index.IndexEngine.get_loc
  File "pandas/_libs/index.pyx", line 138, in pandas._libs.index.IndexEngine.get_loc
  File "pandas/_libs/hashtable_class_helper.pxi", line 1619, in pandas._libs.hashtable.PyObjectHashTable.get_item
  File "pandas/_libs/hashtable_class_helper.pxi", line 1627, in pandas._libs.hashtable.PyObjectHashTable.get_item
KeyError: 'GOI_score'
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/opt/pyenv/tobias-env/bin/TOBIAS", line 11, in <module>
    load_entry_point('tobias==0.11.1.dev0', 'console_scripts', 'TOBIAS')()
  File "/opt/pyenv/tobias-env/lib/python3.8/site-packages/tobias-0.11.1.dev0-py3.8-linux-x86_64.egg/tobias/TOBIAS.py", line 154, in main
  File "/opt/pyenv/tobias-env/lib/python3.8/site-packages/tobias-0.11.1.dev0-py3.8-linux-x86_64.egg/tobias/tools/bindetect.py", line 526, in run_bindetect
  File "/opt/pyenv/tobias-env/lib/python3.8/site-packages/tobias-0.11.1.dev0-py3.8-linux-x86_64.egg/tobias/tools/bindetect.py", line 526, in <listcomp>
  File "/opt/python-3.8.2/lib/python3.8/multiprocessing/pool.py", line 768, in get
    raise self._value
KeyError: 'GOI_score'

```

I was wondering if there is a way to set up the p-value cut off after the BINDetect runs but I was unable to understand if this p-value information is kept in the output files, e.g. the _bound.bed files. 
Thank you.

Danvor

**Comment by msbentsen** [comment]
Hi Danvor, 

Do you get the same error when using the test-data? Such as using:
`TOBIAS BINDetect --motifs test_data/motifs.jaspar --signals test_data/Bcell_footprints.bw --genome test_data/genome.fa.gz --peaks test_data/merged_peaks_annotated.bed --peak_header test_data/merged_peaks_annotated_header.txt --outdir BINDetect_single_output --cond_names Bcell --cores 8` 

What I notice is that you have very few input regions (7). Because the background is estimated from these regions, there might not be enough to properly estimate the bound/unbound threshold during the run. I will have a look at the error though.

The p-value is used to set the threshold of bound/unbound sites during the run, but if you want, you can use the "_score" columns to set another threshold after the run. During the run, this is estimated at:
```
2020-05-04 16:08:47 (1329) [INFO]	Estimating bound/unbound threshold
2020-05-04 16:08:47 (1329) [STATS]	- Threshold estimated at: 3.53463
```
If you look into the _overview.txt files in the output, you can make your own split with another value from the _score columns.

Best Mette

**Comment by Danvor** [comment]
Hi, 
I do get the same error using the test data under similar conditions as what I have posted, i.e. with 7 peaks from the test peak set. However, when I run the command you suggested everything works fine for the default p-value, and, indeed, when I run the BINDetect on my data with genome wide peak set everything works fine too with the default p-value.
Thanks.

**Comment by msbentsen** [comment]
I found the reason for this issue - when the number of peaks is very low (or the p-value threshold for finding TFBS is very strict), some TF's have 0 output sites. This broke the writing of the output files, but it should be fixed now in `TOBIAS 0.11.2`.

However, I would caution you against running BINDetect with such a low peak number, as the program uses the whole peak set to estimate the bound/unbound threshold. Even if you are only interested in some peaks, it is better to set `--peaks` to all peaks, and then subset the sites afterwards. This way, you make sure that the bound/unbound threshold is meaningful in the context of the whole genome. I also added a warning from BINDetect for this case.
